pkg/controller/jobs/pod/event_handlers.go

```diff
@@ -0,0 +1,144 @@
+		return reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Namespace: p.Namespace,
+				Name:      p.Name,
+			},
+		}
+	}
+	return reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      groupName,
+			Namespace: fmt.Sprintf("group/%s", p.Namespace),
+		},
+	}
+}
+
+// podEventHandler will convert reconcile requests for pods in group from "<namespace>/<pod-name>" to
+// "group/<namespace>/<group-name>".
+type podEventHandler struct {
+	client client.Client
+}
+
+func (h *podEventHandler) Create(ctx context.Context, e event.CreateEvent, q workqueue.RateLimitingInterface) {
+	h.queueReconcileForPod(ctx, e.Object, q)
+}
+
+func (h *podEventHandler) Update(ctx context.Context, e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	h.queueReconcileForPod(ctx, e.ObjectNew, q)
+}
+
+func (h *podEventHandler) Delete(ctx context.Context, e event.DeleteEvent, q workqueue.RateLimitingInterface) {
+	h.queueReconcileForPod(ctx, e.Object, q)
+}
+
+func (h *podEventHandler) Generic(ctx context.Context, e event.GenericEvent, q workqueue.RateLimitingInterface) {
+	h.queueReconcileForPod(ctx, e.Object, q)
+}
+
+func (h *podEventHandler) queueReconcileForPod(ctx context.Context, object client.Object, q workqueue.RateLimitingInterface) {
+	p, ok := object.(*corev1.Pod)
+	if !ok {
+		return
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("pod", klog.KObj(p))
+	log.V(5).Info("Queueing reconcile for pod")
+
+	q.Add(reconcileRequestForPod(p))
+}
+
+type parentWorkloadHandler struct {
+	client client.Client
+}
+
+func (h *parentWorkloadHandler) Create(ctx context.Context, e event.CreateEvent, q workqueue.RateLimitingInterface) {
+	h.queueReconcileForChildPod(ctx, e.Object, q)
+}
+
+func (h *parentWorkloadHandler) Update(ctx context.Context, e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	h.queueReconcileForChildPod(ctx, e.ObjectNew, q)
+}
+
+func (h *parentWorkloadHandler) Delete(context.Context, event.DeleteEvent, workqueue.RateLimitingInterface) {
+}
+
+func (h *parentWorkloadHandler) Generic(ctx context.Context, e event.GenericEvent, q workqueue.RateLimitingInterface) {
+}
+
+func (h *parentWorkloadHandler) queueReconcileForChildPod(ctx context.Context, object client.Object, q workqueue.RateLimitingInterface) {
+	w, ok := object.(*kueue.Workload)
+	if !ok {
+		return
+	}
+	log := ctrl.LoggerFrom(ctx).WithValues("workload", klog.KObj(w))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	if len(w.ObjectMeta.OwnerReferences) == 0 {
+		return
+	}
+	log.V(5).Info("Queueing reconcile for parent pods")
+
+	for _, ownerRef := range w.ObjectMeta.OwnerReferences {
+		var parentPod corev1.Pod
+
+		err := h.client.Get(ctx, types.NamespacedName{Name: ownerRef.Name, Namespace: w.ObjectMeta.Namespace}, &parentPod)
+
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			log.Error(err, "Unable to get parent pod")
+			return
+		}
+
+		if groupName := fromObject(&parentPod).groupName(); groupName == "" {
```

Comment:
avoid calling `fromObject`. Instead, change the method `groupName` to work on `corev1.Pod`. This is to avoid unnecessary allocations.